### PR TITLE
RavenDB-12024

### PR DESCRIPTION
**Pull request by garayx**
WIP DO NOT MERGE

### CODE REVIEW

**Comment by garayx** [comment]
WIP DO NOT MERGE

File: test/SlowTests/Server/Documents/Indexing/Debugging/RavenDB_5577.cs [hunk 1]
```diff
@@ -96,7 +97,7 @@ select new
             }
         }
 
-        [Theory]
+        [TheoryAndSkipWhen32BitsEnvironment]
```
Comment (ppekrol) [inline test/SlowTests/Server/Documents/Indexing/Debugging/RavenDB_5577.cs]: Rename to `Theory64Bit`

File: test/SlowTests/Issues/RavenDB_11904.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@ namespace SlowTests.Issues
 {
     public class RavenDB_11904 : RavenTestBase
     {
-        [Fact]
+        [FactAndSkipWhen32BitsEnvironment]
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_11904.cs]: Rename to: `Fact64Bit`

File: test/SlowTests/Issues/RavenDB_10382.cs [hunk 3]
```diff
@@ -22,14 +22,27 @@ public void ShouldWork()
             var random = new Random(1);
             var bytes = new byte[1024 * 8];
 
+            int treeKeys;
+            int treeKeysToDel;
+
+            if (IntPtr.Size == 8)
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_10382.cs]: Avoid doing that.

Create two separate test cases:
1. `ShouldWork32` with `Fact32Bit`
2. `ShouldWork` with `Fact64Bit`

File: test/StressTests/Client/Attachments/AttachmentsSessionAsyncStress.cs [hunk 4]
```diff
@@ -13,6 +14,8 @@ public class AttachmentsSessionAsyncStress : NoDisposalNeeded
         [InlineData(1_000_000)]
         public async Task PutLotOfAttachments(int count)
         {
+            if (IntPtr.Size == sizeof(int))
```
Comment (ppekrol) [inline test/StressTests/Client/Attachments/AttachmentsSessionAsyncStress.cs]: Avoid doing that. Create a separate method for 32Bit

File: test/StressTests/Client/Attachments/AttachmentsSessionStress.cs [hunk 5]
```diff
@@ -46,6 +47,8 @@ public void StressPutLotOfAttachments(int count)
         [InlineData(100_000)]
         public async Task StressPutLotOfAttachmentsAsync(int count)
         {
+            if (IntPtr.Size == sizeof(int))
```
Comment (ppekrol) [inline test/StressTests/Client/Attachments/AttachmentsSessionStress.cs]: Avoid doing that. Create a separate method for 32Bit

File: test/StressTests/Databases.cs [hunk 6]
```diff
@@ -26,6 +26,8 @@ public void CanHandleMultipleDatabasesOnWrite(int numberOfDatabases)
 
             using (var store = GetDocumentStore())
             {
+                if (IntPtr.Size == sizeof(int))
```
Comment (ppekrol) [inline test/StressTests/Databases.cs]: Avoid doing that. Create a separate method for 32Bit

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Voron/Impl/Paging/Windows32BitsMemoryMapPager.cs [hunk 7]
```diff
@@ -166,7 +166,7 @@ public override bool EnsureMapped(IPagerLevelTransactionState tx, long pageNumbe
             }
 
             var ammountToMapInBytes = NearestSizeToAllocationGranularity((distanceFromStart + numberOfPages) * Constants.Storage.PageSize);
-            MapPages(state, allocationStartPosition, ammountToMapInBytes);
+            MapPages(state, allocationStartPosition, ammountToMapInBytes, true);
```
Comment (ayende) [inline src/Voron/Impl/Paging/Windows32BitsMemoryMapPager.cs]: Is there a reason this will _ever_ be false?

File: test/SlowTests/Issues/RavenDB_10382.cs (hunk 3 above)
Comment (ayende) [reply test/SlowTests/Issues/RavenDB_10382.cs]: Or create a `Theory` that has `InlineData32Bits` and `InlineData64Bits`

File: test/SlowTests/Issues/RavenDB_10382.cs [hunk 8]
```diff
@@ -22,14 +23,88 @@ public void ShouldWork()
...
+                        count++;
+                    } while (it.MoveNext());
+
+                    Assert.Equal(treeKeysAssumedLeft, count);
+                }
+            }
+        }
+
+        [Fact32Bit]
+        public void ShouldWork32()
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_10382.cs:20]: You can create `private void ShouldWorkInternal(int treeKeys, int treeKeysToDel)` to avoid duplication

File: test/StressTests/Databases.cs [hunk 9]
```diff
@@ -18,7 +18,7 @@ namespace StressTests
 {
     public class Databases : RavenTestBase
     {
-        [NightlyBuildTheory]
+        [NightlyBuildTheory64]
```
Comment (ppekrol) [inline test/StressTests/Databases.cs]: Let's be consistent. Either go with `<name>32|64` or `<name>32Bit|64Bit` for all attribute names.

File: test/Tests.Infrastructure/Fact32BitAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+        public Fact32BitAttribute()
+        {
+            if (PlatformDetails.Is32Bits)
+            {
+                Is32Bit = true;
+                return;
+            }
+            if (bool.TryParse(Environment.GetEnvironmentVariable("VORON_INTERNAL_ForceUsing32BitsPager"), out var result))
+                if (result == false)
```
Comment (ppekrol) [inline test/Tests.Infrastructure/Fact32BitAttribute.cs]: Is this correct? Shouldn't be the opposite?

File: test/Tests.Infrastructure/NightlyBuildTheory32Attribute.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+        public NightlyBuildTheory32Attribute()
+        {
+            if (PlatformDetails.Is32Bits)
+            {
+                Is32Bit = true;
+                return;
+            }
+            if (bool.TryParse(Environment.GetEnvironmentVariable("VORON_INTERNAL_ForceUsing32BitsPager"), out var result))
+                if (result == false)
```
Comment (ppekrol) [inline test/Tests.Infrastructure/NightlyBuildTheory32Attribute.cs]: Opposite?

File: test/Tests.Infrastructure/Theory32BitAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+        public Theory32BitAttribute()
+        {
+            if (PlatformDetails.Is32Bits)
+            {
+                Is32Bit = true;
+                return;
+            }
+            if (bool.TryParse(Environment.GetEnvironmentVariable("VORON_INTERNAL_ForceUsing32BitsPager"), out var result))
+                if (result == false)
```
Comment (ppekrol) [inline test/Tests.Infrastructure/Theory32BitAttribute.cs]: Opposite?

File: src/Voron/Data/BTrees/Tree.Compressed.cs [hunk 13]
```diff
@@ -47,7 +47,7 @@ private bool TryCompressPageNodes(Slice key, int len, TreePage page)
                     {
                         // we've just put a decompressed page to the cache however we aren't going to compress it
                         // need to invalidate it from the cache
-                        DecompressionsCache.Invalidate(page.PageNumber, DecompressionUsage.Write);
+                        DecompressionsCache?.Invalidate(page.PageNumber, DecompressionUsage.Write);
```
Comment (arekpalinski) [inline src/Voron/Data/BTrees/Tree.Compressed.cs]: Why did you need to use `?.` operator in all those places? If tree is compressed then we are guaranteed that `DecompressionsCache` isn't null

File: src/Voron/Data/BTrees/Tree.Compressed.cs [hunk 14]
```diff
@@ -303,32 +303,40 @@ private void DeleteOnCompressedPage(TreePage page, Slice keyToDelete, ref TreeCu
...
                 }
 
-                if (page.LastMatch != 0)
-                    return null;
+                return new DecompressedReadResult(GetValueReaderFromHeader(node), decompressed);
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
+                throw;
```
Comment (arekpalinski) [inline src/Voron/Data/BTrees/Tree.Compressed.cs]: Probably debug code. Should be removed.

File: src/Raven.Server/Documents/Indexes/Static/DynamicBlittableJson.cs [hunk 15]
```diff
@@ -35,7 +36,8 @@ static DynamicBlittableJson()
                 new CompareKey(Constants.Documents.Metadata.Id, 1),
                 new CompareKey(Constants.Documents.Metadata.ChangeVector, 1),
                 new CompareKey(Constants.Documents.Metadata.LastModified, 1),
-                new CompareKey("Count", 2)
+                new CompareKey("Count", 2),
+                new CompareKey("@etag", 1),
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Static/DynamicBlittableJson.cs]: Use Constants here

File: src/Raven.Server/Documents/Indexes/IndexStorage.cs [hunk 16]
```diff
@@ -536,12 +536,12 @@ public unsafe void WriteReferences(CurrentIndexingScope indexingScope, RavenTran
...
                         {
                             var etag = collections.Value;
-
                             var result = collectionEtagTree.Read(collectionKey);
                             var oldEtag = result?.Reader.ReadLittleEndianInt64();
                             if (oldEtag >= etag)
                                 continue;
-
+                            if (oldEtag < etag)
+                                etag = oldEtag.Value;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/IndexStorage.cs]: Not sure that I follow?

This is always 'true', I think that you will never increase the value, so you it will always process the same references, all over again.

Comment (garayx) [reply src/Raven.Server/Documents/Indexes/IndexStorage.cs]: Its always oldEtag = etag on 64bit and on 32bit 
but if the index batch is splitted (because of the 8mb limitation on 32bit) then oldEtag is the last etag of the document which will be indexed next, but if there no check for (oldEtag < etag) then etag just overwrite oldEtag and we miss indexing few documents. However I think I should add a check for DoMoreWork == true boolean.
if(DoMoreWork &&  oldEtag < etag)

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/IndexStorage.cs]: Let's add a comment here

Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/IndexStorage.cs]: Fix here: https://github.com/ravendb/ravendb/pull/7728

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs [hunk 17]
```diff
@@ -134,33 +131,22 @@ private unsafe bool HandleDocuments(ActionType actionType, DocumentsOperationCon
...
-                                            });
+                                            .Select(tombstone => new Reference { Etag = tombstone.Etag, Key = tombstone.LowerId });
                                         break;
                                     default:
                                         throw new NotSupportedException();
                                 }
 
-                                foreach (var referencedDocument in references)
+                                var list = references.ToList();
+                                foreach (var referencedDocument in list)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: Why this is needed? Looks like a debugging code?

Comment (garayx) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: done

File: src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs [hunk 18]
```diff
@@ -23,8 +24,6 @@ public class HandleReferences : IIndexingWork
         private readonly DocumentsStorage _documentsStorage;
         private readonly IndexStorage _indexStorage;
 
-        private readonly Reference _reference = new Reference();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: Does it need to be removed?

Comment (garayx) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: image

new Reference used in lambda expression instead (line 134 and 142)

Comment (ppekrol) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: I understood that, but why we can't have one instance? When enumerating without a `.ToList()` that you removed it should work as it should.

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Workers/HandleReferences.cs]: It was debug code, IIRC, and we can change it back, I think

**Comment by ppekrol** [comment]
Rebased and fixed the HandleReferences and IndexStorage

**Review by ayende — APPROVED** [review]

**Review by ppekrol — APPROVED** [review]
